http-client/Network/HTTP/Client/Connection.hs

```diff
@@ -184,10 +189,29 @@ openSocket tweakSocket addr =
             NS.connect sock (NS.addrAddress addr)
             return sock)
 
+-- Pick up an IP using an approximation of the happy-eyeballs algorithm:
+-- https://datatracker.ietf.org/doc/html/rfc8305
+-- 
 firstSuccessful :: [NS.AddrInfo] -> (NS.AddrInfo -> IO a) -> IO a
-firstSuccessful []     _  = error "getAddrInfo returned empty list"
-firstSuccessful (a:as) cb =
-    cb a `E.catch` \(e :: E.IOException) ->
-        case as of
-            [] -> E.throwIO e
-            _  -> firstSuccessful as cb
+firstSuccessful []        _  = error "getAddrInfo returned empty list"
+firstSuccessful addresses cb = do
+    result <- newEmptyMVar
+    either E.throwIO pure =<<
+        withAsync (tryAddresses result)
+            (\_ -> takeMVar result)
+  where
+    -- https://datatracker.ietf.org/doc/html/rfc8305#section-5
+    connectionAttemptDelay = 250 * 1000
+
+    tryAddresses result = do
+        z <- forConcurrently (zip addresses [0..]) $ \(addr, n) -> do
+            when (n > 0) $ threadDelay $ n * connectionAttemptDelay
+            tryAddress addr
+        
+        let lastException = take 1 $ reverse $ filter isLeft z
+        for_ lastException $ tryPutMVar result
```

Comment:
I know this is immune from a deadlock based on the first clause checking for an empty list. However, I think it would still be better to have an explicit pattern match here and generate a new exception in the case of a `Nothing`.

Also, I'm not sure that filtering for `isLeft` is an improvement. By construction, there should be no `Right` values in the list. Instead of filtering, _if_ I was going to analyze the values, I think I'd rather see an `assert` to guarantee the value is `Left`.

Reply:
I pushed a version that does pretty much that, pattern matching and calling `error` if unexpected happens.